### PR TITLE
Use event.code and not event.key for wasdqe controls so this works for non QWERTY keyboards

### PR DESCRIPTION
**Pull request by vincentfretin**
This fixes https://github.com/mozilla/hubs/issues/2397

### CODE REVIEW

File: src/systems/userinput/devices/keyboard.js [hunk 1]
```diff
@@ -45,7 +55,10 @@ export class KeyboardDevice {
         this.keys = {};
         this.seenKeys.clear();
       } else {
-        const key = event.key.toLowerCase();
+        let key = event.key.toLowerCase();
+        // Use event.code and not event.key for wasdqe controls so this works
+        // for non QWERTY keyboards.
+        key = (event.code && CODE_TO_KEY[event.code]) || key;
```
Comment (Slayug) [inline src/systems/userinput/devices/keyboard.js:61]:
```suggestion
        // Use event.code and not event.key for wasdqe controls so this works
        // for non QWERTY keyboards.
        const key = CODE_TO_KEY[event.code] ?? event.key.toLowerCase();
```
`event.code` always truthy in modern browser, so `key = CODE_TO_KEY[event.code] || event.key.toLowerCase();`
`event.code` always falsy in old browser, so `key = event.key.toLowerCase();`

You can support backward compatibility with:
```
export const KEYCODE_TO_KEY = {
  65: "q",
  68: "d",
  69: "e",
  81: "a",
  83: "s",
  90: "w"
};
// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/code#browser_compatibility
// https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode#browser_compatibility
const key = CODE_TO_KEY[event.code] ?? KEYCODE_TO_KEY[event.keycode] ?? event.key.toLowerCase()
```

**Review by Slayug — CHANGES_REQUESTED** [review]
Hi, short code and backward compatibility suggestions

File: src/systems/userinput/devices/keyboard.js (hunk 1 above)
Comment (vincentfretin) [reply src/systems/userinput/devices/keyboard.js:61]: You are mixing two different things, code "KeyW", and keycode "90". There is no need for a mapping between keycode to key here, we already have to key in `event.key`.

"event.code always truthy in modern browser, so key = CODE_TO_KEY[event.code] || event.key.toLowerCase();
event.code always falsy in old browser, so key = event.key.toLowerCase();"
Yes this is exactly what I want. For old browsers, they don't have the required thing to know which key is pressed (event.code), then we default to the wrong qwerty behavior, no choice.

Comment (vincentfretin) [reply src/systems/userinput/devices/keyboard.js:61]: For your other way to write it
`const key = CODE_TO_KEY[event.code] ?? event.key.toLowerCase();`
This should work I guess yes.
I'll leave this decision to Mozilla if they decide to include a fix for this issue. I won't spend time to modify this PR.

Comment (Slayug) [reply src/systems/userinput/devices/keyboard.js:61]: I've test it works and more clean for the final JS code. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator

Yes you are right for the backward thing.

**Comment by netpro2k** [comment]
One thing to note with this solution is that on some layouts some key mappings may be completely masked, where the physical hardcoded keys here overlap with a key mapped by character elsewhere. To fix this we probably need to switch all mappings to be based on physical location rather than character. We also eventually would want to let the users customize these keys and the problem largely goes away.

This also means that for non QWERTY users, any of the tutorial tips and documentation are going to be incorrect. The tutorial tips are fixable using the experimental `navigator.keyboard.getLayoutMap()` API, but this is currently only supported in Chrome. The documentation is currently static, so we would probably need to do some work to get arbitrary javascript onto those pages to dynamically explain the key mappings... Instead we may want to just show an image of the "reference keyboard" so people can just compare it to their physical layout https://www.w3.org/TR/uievents-code/#key-alphanumeric-writing-system. 

I have added a fix for the in client tutorial tips at least https://github.com/mozilla/hubs/commit/b99a0c91463d0ccacfbd2565dc7dc6cd5a51cda9.

Even given these tradeoffs this solution feels like its better than the status queue, and probably provides an overall better experience for users of non QWERTY layouts. Keys that it might mask are certainly less important than the primary movement keys.

I am going to merge it with my patch and a TODO to improve it later. Thanks for the contribution @vincentfretin, sorry again we are so behind on accepting PRs.

**Pull request merged**